src/validation.h

```diff
@@ -676,7 +676,13 @@ class CChainState
     void UnloadBlockIndex() EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
 
     /** Check whether we are doing an initial block download (synchronizing from disk or network) */
-    bool IsInitialBlockDownload() const;
+    bool IsInitialBlockDownload() const EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
+
+    /**
+     * Same as `IsInitialBlockDownload()`, but also locks mutex cs_main during
+     * initial block download. Use for callers not already holding the lock.
+     */
+    bool IsIBD() const;
```

Comment:
should we add a `LOCKS_EXCLUDED` to prevent accidentally calling this when we know we have cs_main locked?

Reply:
You read my mind. Added an `LOCK_EXCLUDED` thread safety annotation to the declaration along with an `AssertLockNotHeld(::cs_main);` run-time assertion in the definition.